engine: remove TempDir, introduce AuxiliaryDir

tbg opened the pull request:
This is in preparation for Raft sideloading (#16263). We need a permanent
location to link files into RocksDB from. For temporary files, we can simply
use a subdirectory that we wipe on node start.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/16605)
<!-- Reviewable:end -->

danhhz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Doesn't have to be addressed here, but we should probably revisit the in-memory RocksDB story. Currently, if you put a file in the auxiliary dir returned by an in-memory RocksDB and try to link it in, it will fail (because it's in the os temp dir, not the in-memory env)

---

Review status: 0 of 7 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16605#-:-Kn0iHnWvDVfi9EkC0Dt:bz4pn0s)*
<!-- Sent from Reviewable.io -->

tbg committed:
engine: remove TempDir, introduce AuxiliaryDir

This is in preparation for Raft sideloading (#16263). We need a permanent
location to link files into RocksDB from. For temporary files, we can simply
use a subdirectory that we wipe on node start.

tschottdorf commented:
TFTR! Yeah, I'm not clear how this linking in business is going to work with in-memory engines. Let's chat about this at some point.
